File: examples/word-run-formatting.ts
/**
 * 例子（Epic-38）：演示 Run 格式访问器（bold/italic/underline/fontSize/color）。
 *
 * 跑法：
 *   bun run examples/word-run-formatting.ts <output.docx>
 */

import { Paragraph, Run, Text, WordprocessingDocument } from "../src/word/index.js";

async function main(): Promise<void> {
  const [outputPath] = process.argv.slice(2);
  if (outputPath === undefined) {
    process.stderr.write("usage: word-run-formatting <output.docx>\n");
    process.exit(2);
  }

  const doc = WordprocessingDocument.create();
  const body = doc.mainDocumentPart?.document.firstChild()!;
  const append = (e: unknown): void =>
    (body as { appendChild: (e: unknown) => void }).appendChild(e);

  function makeStyledRun(
    text: string,
    opts: {
      bold?: boolean;
      italic?: boolean;
      underline?: string;
      fontSizeHalfPoints?: number;
      colorHex?: string;
    } = {},
  ): Run {
    const r = new Run();
    const t = new Text();
    t.text = text;
    r.appendChild(t);
    if (opts.bold !== undefined) r.bold = opts.bold;
    if (opts.italic !== undefined) r.italic = opts.italic;
    if (opts.underline !== undefined) r.underline = opts.underline;
    if (opts.fontSizeHalfPoints !== undefined) r.fontSizeHalfPoints = opts.fontSizeHalfPoints;
    if (opts.colorHex !== undefined) r.colorHex = opts.colorHex;
    return r;
  }

  function addLine(...runs: Run[]): void {
    const p = new Paragraph();
    for (const r of runs) p.appendChild(r);
    append(p);
  }

  addLine(makeStyledRun("Plain text. "), makeStyledRun("Bold.", { bold: true }));
  addLine(makeStyledRun("Italic.", { italic: true }));
  addLine(makeStyledRun("Underlined single.", { underline: "single" }));
  addLine(makeStyledRun("Underlined double.", { underline: "double" }));
  addLine(makeStyledRun("Big red.", { fontSizeHalfPoints: 48, colorHex: "FF0000", bold: true }));
  addLine(
    makeStyledRun("Combo: bold italic underlined 14pt blue.", {
      bold: true,
      italic: true,
      underline: "single",
      fontSizeHalfPoints: 28,
      colorHex: "0000FF",
    }),
  );

  await doc.saveAsAsync(outputPath);
  process.stdout.write(`Wrote ${outputPath} (run format showcase)\n`);
}

main().catch((err) => {
  process.stderr.write(`failed: ${(err as Error).stack ?? err}\n`);
  process.exit(1);
});

File: tools/cross-sdk-verify/replicas/WordRunFormatting.cs
// Replica of examples/word-run-formatting.ts
// Creates a docx showcasing run-level formatting: bold, italic, underline, fontSize, color.

using DocumentFormat.OpenXml;
using DocumentFormat.OpenXml.Packaging;
using DocumentFormat.OpenXml.Wordprocessing;

namespace CrossSdkVerify.Replicas;

public static class WordRunFormatting
{
    public static void Run(string outputPath)
    {
        using var doc = WordprocessingDocument.Create(outputPath,
            DocumentFormat.OpenXml.WordprocessingDocumentType.Document);

        var mainPart = doc.AddMainDocumentPart();
        mainPart.Document = new Document();
        var body = new Body();
        mainPart.Document.Append(body);

        // Paragraph 1: "Plain text. " + "Bold."
        body.Append(MakePara(
            MakeRun("Plain text. "),
            MakeRun("Bold.", bold: true)
        ));

        // Paragraph 2: "Italic."
        body.Append(MakePara(MakeRun("Italic.", italic: true)));

        // Paragraph 3: "Underlined single."
        body.Append(MakePara(MakeRun("Underlined single.", underline: "single")));

        // Paragraph 4: "Underlined double."
        body.Append(MakePara(MakeRun("Underlined double.", underline: "double")));

        // Paragraph 5: "Big red." — 24pt (48 half-points), FF0000, bold
        body.Append(MakePara(MakeRun("Big red.", fontSizeHalfPt: 48, colorHex: "FF0000", bold: true)));

        // Paragraph 6: combo
        body.Append(MakePara(MakeRun(
            "Combo: bold italic underlined 14pt blue.",
            bold: true, italic: true, underline: "single",
            fontSizeHalfPt: 28, colorHex: "0000FF")));

        body.Append(new SectionProperties());
        mainPart.Document.Save();
    }

    private static Paragraph MakePara(params Run[] runs)
        => new Paragraph(runs);

    private static Run MakeRun(
        string text,
        bool? bold = null,
        bool? italic = null,
        string? underline = null,
        int? fontSizeHalfPt = null,
        string? colorHex = null)
    {
        var rPr = new RunProperties();
        if (bold == true) rPr.Append(new Bold());
        if (italic == true) rPr.Append(new Italic());
        if (underline != null)
        {
            var uv = underline == "single" ? UnderlineValues.Single : UnderlineValues.Double;
            rPr.Append(new Underline { Val = uv });
        }
        if (fontSizeHalfPt != null)
            rPr.Append(new FontSize { Val = fontSizeHalfPt.ToString() });
        if (colorHex != null)
            rPr.Append(new Color { Val = colorHex });

        var run = new Run(new Text(text) { Space = SpaceProcessingModeValues.Preserve });
        if (rPr.HasChildren)
            run.PrependChild(rPr);
        return run;
    }
}
